test/jdk/tools/sincechecker/SinceChecker.java

```diff
@@ -0,0 +1,828 @@
+    }
+
+    private void processClassElement(TypeElement te, String version, Types types, Elements elements) {
+        if (!isDocumented(te)) {
+            return;
+        }
+        persistElement(te.getEnclosingElement(), te, types, version);
+        elements.getAllMembers(te).stream()
+                .filter(this::isDocumented)
+                .filter(this::isMember)
+                .forEach(element -> persistElement(te, element, types, version));
+        te.getEnclosedElements().stream()
+                .filter(element -> element.getKind().isDeclaredType())
+                .map(TypeElement.class::cast)
+                .forEach(nestedClass -> processClassElement(nestedClass, version, types, elements));
+    }
+
+    public void persistElement(Element explicitOwner, Element element, Types types, String version) {
+        String uniqueId = getElementName(explicitOwner, element, types);
+        IntroducedIn introduced = classDictionary.computeIfAbsent(uniqueId, i -> new IntroducedIn());
+        if (isPreview(element, uniqueId, version)) {
+            if (introduced.introducedPreview == null) {
+                introduced.introducedPreview = version;
+            }
+        } else {
+            if (introduced.introducedStable == null) {
+                introduced.introducedStable = version;
+            }
+        }
+    }
+
+    private boolean isPreview(Element el, String uniqueId, String currentVersion) {
+        while (el != null) {
+            Symbol s = (Symbol) el;
+            if ((s.flags() & Flags.PREVIEW_API) != 0) {
+                return true;
+            }
+            el = el.getEnclosingElement();
+        }
+
+        return LEGACY_PREVIEW_METHODS.containsKey(currentVersion)
+                && LEGACY_PREVIEW_METHODS.get(currentVersion).contains(uniqueId);
+    }
+
+    void error(String message) {
+        System.err.println(message);
+        errorCount++;
+    }
+
+    private void checkModule(String moduleName) throws Exception {
+        Path home = Paths.get(System.getProperty("java.home"));
+        Path srcZip = home.resolve("lib").resolve("src.zip");
+        if (Files.notExists(srcZip)) {
+            //possibly running over an exploded JDK build, attempt to find a
+            //co-located full JDK image with src.zip:
+            Path testJdk = Paths.get(System.getProperty("test.jdk"));
+            srcZip = testJdk.getParent().resolve("images").resolve("jdk").resolve("lib").resolve("src.zip");
+        }
+        if (!Files.exists(srcZip) && !Files.isDirectory(srcZip)) {
+            throw new SkippedException("Skipping Test because src.zip wasn't found");
+        }
+        if (Files.isReadable(srcZip)) {
+            URI uri = URI.create("jar:" + srcZip.toUri());
+            try (FileSystem zipFO = FileSystems.newFileSystem(uri, Collections.emptyMap())) {
+                Path root = zipFO.getRootDirectories().iterator().next();
+                Path packagePath = root.resolve(moduleName);
```

Comment:
Here, and infuses of this name, including in method signatures, it is strange to see the variable named `packagePath` since it is the path of the unnamed package in the module.

`modulePath` would likely be a better name, unless you are worried about possible confusion with the `--module-path` option, in which case maybe `moduleDir` or `moduleDirectory` would work.